src/s_inter.c

```diff
@@ -138,6 +171,472 @@ void sys_set_temppath(void);
+    if(ret)
+    {
+        errno = EPERM;
+        perror("Error while reading from ring_buffer in rb_recv");
+        return -1;
+    }
+    if(msglen > buflen) {
+        if(rbskt->rs_preserve_boundaries) {
+            printf("buflen: %zu, msglen: %zd, actualLength: %d\n", buflen, msglen, actualLength);
+            printf("warning: incoming packet truncated from %d to %d bytes.", ret, INBUFSIZE);
+            // drain buffer
+            char dest;
+            while(1 == rb_read_from_buffer(rb, &dest, 1))
+                    ;
+        }
+    }
+    return actualLength;
+}
+
+// if this is called with (1), then the audio thread will not manage the I/O
+// and the user will be responsible to regularly call sys_doio()
+void sys_dontmanageio(int status)
+{
+    pd_this->pd_inter->i_dontmanageio = status;
+}
+
+// register a callback to delete an object when a send fd fails in the IO
+// thread. This allows to report errors asynchronously
+void sys_addsendfdrmfn(int fd, t_fdsendrmfn fn, void* ptr)
+{
+    int nfd, size;
+    t_fdsend *fs;
+    if (!pd_this->pd_inter->i_fdsend)
+    {
+        /* create an empty FD list */
+        pd_this->pd_inter->i_fdsend = (t_fdsend *)t_getbytes(0);
+        pd_this->pd_inter->i_nfdsend= 0;
+    }
+    if(rbrmfdsend_init())
+        return;
+    nfd = pd_this->pd_inter->i_nfdsend;
+    size = nfd * sizeof(t_fdsend);
+    pd_this->pd_inter->i_fdsend = (t_fdsend *)t_resizebytes(
+        pd_this->pd_inter->i_fdsend, size, size + sizeof(t_fdsend));
+    fs = pd_this->pd_inter->i_fdsend + nfd;
+    fs->fds_fd = fd;
+    fs->fds_fn = fn;
+    fs->fds_ptr = ptr;
+    pd_this->pd_inter->i_nfdsend = nfd + 1;
+}
+
+// a partial replacement for the functionalities of sys_domicrosleep() when
+// enabling THREADED_IO. Namely:
+// - check for ready file descriptors, and for each ready one:
+//  - if it has a RT-friendly callback, retrieve the new data from the
+//  socket and place it in a ring buffer. sys_domicrosleep() will then call the
+//  callback which will retrieve the data from the buffer
+//  - otherwise, mark it so that the callback will be called from
+//  sys_domicrosleep()) at the first occasion
+//
+// As a consequence of this:
+//  - for callbacks that are RT-friendly, sys_domicrosleep() does not need to select()
+//  - for callbacks that are not RT-friendly, sys_domicrosleep() doesn't have to select()
+//  at every call, but only when data is actually available.
+static void poll_fds()
+{
+    const unsigned int maxRecv = INBUFSIZE;
+    char* buf = NULL;
+    ssize_t ret;
+    struct timeval timout = {0}; // making this timeout longer would be more efficient (by avoiding spurious wakeups), but it would needlessly postpone writes
+    int pollem = 1;
+    t_fdpoll *fp;
+    if (pollem && pd_this->pd_inter->i_nfdpoll)
+    {
+        fd_set readset;
+        fdp_select(&readset, timout, kFdpManagerIoThread);
+        int count = 0;
+        while((fp = fdp_ready_next(&readset)))
```

Comment:
What's the point of `fdp_ready_next`? Doesn't it basically have quadratic complexity when called in a loop? As a sidenote, it might be a good idea to finally use `poll` instead of `select`, which is much easier to use and has better performance (`select` is actually terrible, because internally it has to scan *all* file descriptors up to the given limit)

Reply:
it was meant as a reusable utility, but it ends up being used in only one place. I was not particularly concerned about it being slightly slower (it's just compares after all) than the previous implementation. It could definitely use an extra state variable to return to the previous complexity, or even be inlined again. I just find that it's easier to read when it is spelled out like this.

Reply:
The problem is that for every iteration `fdp_ready_next` scans through the whole fd list to find the next available fd. On the other hand, we probably don't have hundreds of fds, so in practice it won't matter much. It just caught my eye.

Reply:
yes, that's where an extra state variable that keeps track of where we are in the file list would have helped. But then it would make the code a bit uglier, so maybe I will just restore the previous code until there is an actual need to factor this out for reuse.

Reply:
https://github.com/giuliomoro/pure-data-1/commit/41b58b16a2d2f559e98034cb88db6018b76e739f